crates/hsdev/src/input_file.rs

```diff
@@ -0,0 +1,53 @@
+use std::string::String;
+
+use toml::Value;
+
+pub struct InputData {
+    username: String,
+    password: String,
+    dbname: String,
+    host: String,
+    port: u16,
+}
+
+fn get_str_or_default(toml: &Value, key: &str) -> String {
+    let value = toml.get(key);
+
+    let str = if value.is_none() {
+        eprintln!("Could not read toml field: \"{}\"", key);
+        ""
+    } else {
+        value.unwrap().as_str().unwrap_or_default()
+    };
+
+    String::from(str)
+}
+fn get_int_or_default(toml: &Value, key: &str) -> i64 {
+    let value = toml.get(key);
+
+    if value.is_none() {
+        eprintln!("Could not read toml field: \"{}\"", key);
+        0
+    } else {
+        value.unwrap().as_integer().unwrap_or_default()
+    }
+}
+
+impl InputData {
+    pub fn read(toml: &Value) -> InputData {
+        InputData {
+            username: get_str_or_default(toml, "username"),
+            password: get_str_or_default(toml, "password"),
+            dbname: get_str_or_default(toml, "dbname"),
+            host: get_str_or_default(toml, "host"),
+            port: get_int_or_default(toml, "port") as u16,
+        }
```

Comment:
You can take advantage of deserialization using the `serde` crate and [`Value::try_into()`](https://docs.rs/toml/0.8.14/toml/value/enum.Value.html#method.try_into): derive `Deserialize` on `InputData`, and update this to be something along the lines of `toml_value::try_into::<InputData>()` (and then perform error handling).

I'd also suggest not using `toml` as a variable name, since it can cause confusion with the crate name itself.

Reply:
I just pushed a few commits to address this and the string literal in the unit test.